src/llmcompressor/transformers/compression/quantization_format.py

```diff
@@ -44,74 +94,26 @@ def infer_quantization_format(
-
-    if save_compressed:
-        is_24_structure = (
-            SparsityStructure(sparsity_structure) == SparsityStructure.TWO_FOUR
-        )
-        is_weight_only = len(input_args) == 0 and len(weight_args) > 0
-
-        if (
-            weight_args[0].num_bits == 4
-            and weight_args[0].type == QuantizationType.FLOAT.value
-        ):
-            return CompressionFormat.nvfp4_pack_quantized
 
-        if is_weight_only:  # w4a16 and w8a16
-            is_valid_pack = all(
-                weight_arg.num_bits in [4, 8]
-                and weight_arg.type == QuantizationType.INT.value
-                for weight_arg in weight_args
-            )
-            if not is_valid_pack:  # packing only valid for int4 and int 8
-                return CompressionFormat.naive_quantized
-            if is_24_structure:
-                for arg in weight_args:
-                    if (
-                        arg.strategy is not QuantizationStrategy.CHANNEL.value
-                        and arg.strategy is not QuantizationStrategy.GROUP.value
-                    ):
-                        # marlin24 kernel only applicable for channel/group quantization
-                        return CompressionFormat.pack_quantized
-                return CompressionFormat.marlin_24
-            return CompressionFormat.pack_quantized
-        else:  # w8a8 float and int
-            if len(weight_args) == 1:
-                if (
-                    weight_args[0].type == QuantizationType.FLOAT.value
-                    and weight_args[0].num_bits == 8
-                ):
-                    return CompressionFormat.float_quantized
-                if weight_args[0].type == QuantizationType.INT.value:
-                    return CompressionFormat.int_quantized
-
-            return CompressionFormat.naive_quantized
-    else:
-        # format will be inferred from config
+    if not save_compressed:
         return None
 
+    if quantization_format:
+        return [quantization_format]
 
-def _get_unique_quant_args(model):
-    """
-    Gets a list of all the unique quantization settings present in model
-    """
-    quant_info_weight = []
-    quant_info_inputs = []
+    unique_formats = []
     for submodule in model.modules():
         if is_module_quantized(submodule):
             weight_scheme = submodule.quantization_scheme.weights
             input_scheme = submodule.quantization_scheme.input_activations
-            if weight_scheme is not None:
-                if weight_scheme not in quant_info_weight:
-                    quant_info_weight.append(weight_scheme)
-            if input_scheme is not None:
-                if input_scheme not in quant_info_inputs:
-                    quant_info_inputs.append(input_scheme)
-
-    return quant_info_weight, quant_info_inputs
+            if weight_scheme is None:
+                continue  # no weight quant - nothing to compress
+            compression_format = _get_quant_method(
```

Comment:
Why are we not inferring from `scheme.format`? Doesn't this override anything the user might have passed?

Reply:
Rather than waiting until the end to do inference, why not infer a format at the source?

```python3
class QuantizationScheme:
    format: Optional[CompressionFormat] = None

    @validate_field("format")
    def validate_format(self, value):
        if self.weights and self.input_activations ...
        
        inferred_format = ...
        if value is not None and value != inferred_format:
            logger.warn_once("Consider using inferred scheme")
            
        return value or inferred_format
```

Reply:
Okay. I'll also mention that this way of doing it also adds unnecessary runtime and redundant work. `submodule.quantization_scheme` is a pointer to a scheme, and this function modifies each submodule.

In effect, this function writes `submodule.quantization_scheme.format` for each module that shares that scheme.